src/dowel/csv_output.py

```diff
@@ -48,7 +50,35 @@ def record(self, data, prefix=''):
                            'Did you change key sets after your first '
                            'logger.log(TabularInput)?'.format(
                                set(self._fieldnames), set(to_csv.keys())))
-
+                
+                # close any existing log file
+                super().close()
+
+                # Read data from the existing log file once
+                with open(self._filename, 'r') as existing_file:
+                    reader = csv.DictReader(existing_file)
+                    file_data = []
+                    for row in reader:
+                        file_data.append(row)
+
+                    # Update header with new keys by union 
+                    self._joint_filednames = set(self._fieldnames)| set(to_csv.keys())
+
+                    # Update the writer with joint headers
+                    self._writer = csv.DictWriter(
```

Comment:
This isn't the only place in the file a `DictWriter` is opened, so it would probably be best to refactor it.

Reply:
Gotcha, `DictWriter` is created before. Will update only the header of the `Dictwriter` when key inconsistency is encountered.